feat: Whisper prompting

connor-henderson opened the pull request:
# What does this PR do?
Closes #22395, thank you @sanchit-gandhi for the descriptive ask!

Note: due to initial scope expansion the commit history includes initial work towards `condition_on_previous_text`, `always_use_initial_prompt`, and pipeline integration, but these efforts have been pushed to a later PR

This this pull request adds 3 new functionalities + tests to support initial prompting functionality within Whisper's `model.generate()` and `tokenizer`: 

- `prompt_ids` param for `model.generate()`:
  -  Optional param of initial prompt ids to provide context for each chunk of text generated by in `model.generate()`
- `get_prompt_ids` Processor method to create initial prompt ids to pass to generate from a passed in string
- Removing the prompt when the tokenizer is decoding if `skip_special_tokens=True`


Example new API usage:
```py
processor = WhisperProcessor.from_pretrained("openai/whisper-tiny")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")
input_features = processor(input_speech, return_tensors="pt").input_features

# --- Without prompt ---
prompt_ids = processor.get_prompt_ids("Leighton")
output_without_prompt = model.generate(input_features)
print(processor.decode(output_without_prompt[0]))
# "<|startoftranscript|><|en|><|transcribe|><|notimestamps|> He has grave doubts whether Sir Frederick Layton's work is really Greek after all and can discover in it but little of Rocky Ithaca.<|endoftext|>"

# --- With prompt ---
prompt_ids = processor.get_prompt_ids("Leighton")
output_with_prompt = model.generate(input_features, prompt_ids=prompt_ids)
print(processor.decode(output_with_prompt[0]))
# "<|startofprev|> Leighton<|startoftranscript|><|en|><|transcribe|><|notimestamps|> He has grave doubts whether Sir Frederick Leighton's work is really Greek after all and can discover in it but little of Rocky Ithaca.<|endoftext|>"
```

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation). **Haven't added anywhere outside of documenting the new generate() arg directly on the function**
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer: @stas00, Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->
@sanchit-gandhi

Commit:
initial working additions

Commit:
clean and rename, add cond stripping initial prompt to decode

Commit:
cleanup, edit create_initial_prompt_ids, add tests

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22496). All of your documentation changes will be reflected on that endpoint.

Commit:
repo consistency, flip order of conditional

Commit:
fix error, move the processor fn to the tokenizer

Commit:
repo consistency, update test ids to corresponding tokenizer

Commit:
use convert_tokens_to_ids not get_vocab...

hollance commented:
Hey this PR looks really good (although I'll leave the actual review to Sanchit or Arthur). 

I was just wondering whether it also makes sense to support the `condition_on_previous_text` option that the OpenAI repo has, since that uses the same mechanism (using the `<|startofprev|>` token). 

In addition, there's [this PR](https://github.com/openai/whisper/pull/1040) that suggests an `always_use_initial_prompt` option that uses the prompt on every segment, not just the first. Might be useful to consider that here as well.

connor-henderson commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1608,6 +1614,18 @@ def generate(
             generation_config.forced_decoder_ids = forced_decoder_ids
 
+        if initial_prompt_ids is not None:
+            if "decoder_start_token_id":
+                raise ValueError(
+                    "When specifying `initial_prompt_ids`, you cannot also specify `decoder_start_token_id` as it gets overwritten."
+                )
```
Replacing this with an actual conditional now. Any idea how the model test I added passed with this?

Commit:
use actual conditional in generate

Commit:
make sytle

hollance commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1519,6 +1520,12 @@ def generate(
                 Whether or not the model is multilingual.
+            initial_prompt_ids (`List[int]`, *optional*):
+                Optional list of token IDs created by passing text to `processor.create_initial_prompt_ids` that is
+                provided as a prompt for the first window. This can be used to provide or "prompt-engineer" a context
```
What kind of window does this refer to? (I'm assuming what's called window here is what we call a chunk. If that's the case we should be consistent with the terminology.)

connor-henderson replied:
The "first window" refers to the initial segment of the input speech, so I believe it can be used interchangeably with chunk yes. Updated the wording to use 'chunk'

hollance commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1608,6 +1615,18 @@ def generate(
+            kwargs.update({"decoder_start_token_id": initial_prompt_ids.pop(0)})
+            initial_prompt_ids.append(generation_config.decoder_start_token_id)
+            initial_prompt_ids.extend([id for _, id in generation_config.forced_decoder_ids])
+            indexed_initial_prompt_ids = [[idx + 1, id] for idx, id in enumerate(initial_prompt_ids)]
```
Nitpick: The `forced_decoder_ids` use a tuple instead of a list, so this should be a tuple too:

```python
indexed_initial_prompt_ids = [(rank + 1, token) for rank, token in enumerate(initial_prompt_ids)]
```
I also used `rank` instead of `idx` and `token` instead of `id`, to be consistent with `tokenizer.get_decoder_prompt_ids`.

hollance replied:
I'm also not 100% sure that `generation_config.forced_decoder_ids` is always filled in here. For example, if we do the following:

```python
forced_decoder_ids = processor.get_decoder_prompt_ids(language="de", task="translate")
predicted_ids = model.generate(input_features, forced_decoder_ids=forced_decoder_ids)
```

then `generation_config.forced_decoder_ids` may not have the `<|de|>` token at this point. In this situation, the correct `forced_decoder_ids` will be filled in by `super().generate(...)`, as far as I can tell.

connor-henderson replied:
Ah good catch! Yes before the implementation was overriding provided forced_decoder_ids, but this should be fixed now. Before it was adding these from the generation_config only, but now checks kwargs first.

hollance commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -586,6 +586,15 @@ def decode(
         Returns:
             `str`: The decoded sentence.
         """
+        initial_prompt_start_id = self.convert_tokens_to_ids("<|startofprev|>")
```
Perhaps this could be an instance variable.

connor-henderson replied:
Now that we've updated this decode function with your above suggestion this line was removed. This same logic now is only used once and its on the processor. Would you still like me to make it an instance var there?

hollance replied:
If it's just used once then it's not really worth making it an instance variable.

hollance commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -714,6 +723,12 @@ def _decode_asr(self, model_outputs, *, return_timestamps, return_language, time
+    def create_initial_prompt_ids(self, text):
+        """Creates an initial prompt for generating text."""
+        intial_prompt_start_id = self.convert_tokens_to_ids("<|startofprev|>")
+        tokenized_text = self(" " + text.strip(), add_special_tokens=False)
```
What is the reason for doing `" " + text.strip()`?

connor-henderson replied:
This is what was done in the [Whisper implementation](https://github.com/openai/whisper/blob/b5851c6c40e753606765ac45b85b298e3ae9e00d/whisper/transcribe.py#L195). I believe the model formatting expects white spaces will be removed from the beginning and at the end of the string, with one space added at the start

hollance commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -1400,6 +1400,28 @@ def test_tiny_specaugment_librispeech(self):
         self.assertTrue(torch.allclose(logits[0][0, 0, :30].cpu(), EXPECTED_LOGITS, atol=1e-4))
 
+    @slow
+    def test_generate_with_intial_prompt(self):
```
Small typo in the function name. :-) But excellent work on adding these unit tests!

hollance reviewed (commented):
Hi @connor-henderson, I was asked by @sanchit-gandhi to do a code review for your PR. It looks pretty good already, just needs a bit of fine-tuning. I'm only just getting familiar with the Whisper code myself, so my opinions don't necessarily always make sense. ;-)

connor-henderson commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -586,6 +586,15 @@ def decode(
             `str`: The decoded sentence.
         """
+        initial_prompt_start_id = self.convert_tokens_to_ids("<|startofprev|>")
+        has_initial_prompt = len(token_ids) > 0 and (token_ids[0] == initial_prompt_start_id)
```
Oh good idea, updated to always skip everything until the bos_token

connor-henderson replied:
Looking into this it appears the `bos_token` is <|endoftext|> unless otherwise set, which we couldn't use for slicing

connor-henderson commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1462,6 +1462,7 @@ def generate(
         task=None,
         language=None,
         is_multilingual=None,
+        initial_prompt_ids: Optional[List[int]] = None,
```
Sounds good, renamed to prompt_ids. 

With regards to the type, I wonder if this is ok here because `prompt_ids` purpose is just to update `forced_decoder_ids` and `decoder_start_token_id` which are type int and List[List[int]] (if I'm not mistaken, saw that [here](https://github.com/huggingface/transformers/blob/48fbd8fa2ebfcf35c14f3bca4055631bf83e7e8b/src/transformers/generation/configuration_utils.py#L178))? If we used torch.Tensor, we would also have to add importing torch to the file with the `get_prompt_ids` function and I don't believe we'd ever require tensor functionality. Just lmk whichever you prefer

connor-henderson commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +91,6 @@ def decode(self, *args, **kwargs):
         """
         return self.tokenizer.decode(*args, **kwargs)
+
+    def create_initial_prompt_ids(self, text):
```
Cool will rename and move it to just being on the processor. My thinking was that users would also want to access the method on the tokenizers directly, but I wasn't sure.

Commit:
initial address comments

connor-henderson commented:
Will pick back up tomorrow on these two before re-requesting review

- [ ] Converting the prompt token to an ID in an instance variable gives an incorrect ID, unlike when its called in decode
- [ ] Bug I found where if the ending text of the prompt matches the start of the transcribed text, that text will not be included in the transcription output. Example: 
<img width="779" alt="Screenshot 2023-04-05 at 1 14 03 AM" src="https://user-images.githubusercontent.com/78612354/229986962-269e4564-2b01-405a-a510-fab7d82c2915.png">

hollance commented:
One more thing we'll need to do, is change the `automatic-speech-recognition` pipeline so that it will actually call `model.generate()` with the prompt, but only for the first chunk (or always if we also decide to support an `always_use_initial_prompt` option). This logic cannot be part of the modeling code, as `model.generate()` has no knowledge of which chunk of audio it's processing.

hollance commented:
I looked a bit more into how this works today, and it turns out that 🤗 Transformers does things a bit differently than the original OpenAI code. 

OpenAI does the following: 

For the first 30-second chunk of audio, it passes the following token sequence to the model's decoder on the first iteration: `<|startofprev|> initial prompt<|startoftranscript|><|en|><|transcribe|>`. And then it decodes the rest of the sequence autoregressively.

Then for the second chunk of audio, it passes the following sequence to the decoder on the first iteration: `<|startofprev|> initial prompt output of the first chunk<|startoftranscript|><|en|><|transcribe|>`. 

For the next chunk, it uses `<|startofprev|> initial prompt output of the first chunk output of the second chunk<|startoftranscript|><|en|><|transcribe|>`

And so on... This list of tokens that it passes in the `<|startofprev|>` section grows longer and longer with each new chunk. 

(When you set the `condition_on_previous_text` option to False, it only uses the output from the previous chunk instead of the complete history. In that case the initial prompt text is only used for the very first chunk.)

Our ASR `pipeline` works quite differently. It also splits up the audio in 30-second chunks but they partially overlap, and then it runs the model on these chunks in parallel. That makes it impossible to pass the previous context to these chunks, as each chunk is processed independently. So we have no way of sending `<|startofprev|> initial prompt output of the first chunk<|startoftranscript|><|en|><|transcribe|>` to the second chunk.

The best we can do is send `<|startofprev|> initial prompt<|startoftranscript|><|en|><|transcribe|>` to the very first chunk only, or always send it to all chunks. So we ignore the "previous context" part and always include the prompt. (The latter would do the same as this open [PR on the OpenAI repo](https://github.com/openai/whisper/pull/1040) for always passing the initial prompt inside `<|startofprev|>` instead of the previous context.)

The suggested modifications to `model.generate()` in this PR make it possible to have both `initial_prompt` and the `condition_on_previous_text` options as in OpenAI, but it would require the user to write their own processing loop to get the same results as OpenAI. So we should definitely continue with this PR, but if we also want to support `initial_prompt` in the `pipeline` we'll have to decide on which approach we want. (It's not possible to have `condition_on_previous_text` in the current pipeline.)

Commit:
initial working add new params to pipeline

Commit:
first draft of sequential generation for condition_on_previous_text

Commit:
add/update tests, make compatible with timestamps

Commit:
make compatible with diff. input kwargs and max length

Commit:
add None check

Commit:
add temperature check

connor-henderson commented:
Ok the additional requested features are now added so I believe this is ready for re-review. Thank you for your comments! 

> However note that this pipeline only processes the first 30 seconds of the audio file. This is fine for audio that is shorter than 30 seconds... In effect, this is the `always_use_initial_prompt` option.

I think I’m missing something here as I’ve tried this on >1 min of audio in the below example where I also added a debug line to decode the tokens inside of the pipeline as they were generated, and it appears to be properly sequential. In any case, if we don’t want this I’ll remove `condition_on_previous_text` from the pipeline just lmk!  
```python
pipe = pipeline(task="automatic-speech-recognition", model="openai/whisper-tiny")
res = pipe(samples, generate_kwargs={ "condition_on_previous_text": True, "prompt_ids": prompt_ids })
# ['<|startofprev|><|startoftranscript|><|en|><|transcribe|><|notimestamps|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel.<|endoftext|>']
# ["<|startofprev|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel.<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Nor is Mr. Quilter's manner less interesting than his matter.<|endoftext|>"]
# ["<|startofprev|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel. Nor is Mr. Quilter's manner less interesting than his matter.<|startoftranscript|><|en|><|transcribe|><|notimestamps|> He tells us that at this festive season of the year with Christmas and roast beef looming before us, similarly drawn from eating and its results occur most readily to the mind.<|endoftext|>"]
# ["<|startofprev|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel. Nor is Mr. Quilter's manner less interesting than his matter. He tells us that at this festive season of the year with Christmas and roast beef looming before us, similarly drawn from eating and its results occur most readily to the mind.<|startoftranscript|><|en|><|transcribe|><|notimestamps|> He has grave doubts whether Sir Frederick Layton's work is really Greek after all and can discover in it but little of Rocky Ithaca.<|endoftext|>"]
# ["<|startofprev|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel. Nor is Mr. Quilter's manner less interesting than his matter. He tells us that at this festive season of the year with Christmas and roast beef looming before us, similarly drawn from eating and its results occur most readily to the mind. He has grave doubts whether Sir Frederick Layton's work is really Greek after all and can discover in it but little of Rocky Ithaca.<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Lennils, pictures are a sort of upguards and atom paintings and Mason's exquisite itals are as national as a jingo poem. Mr. Berkett Foster's landscapes smile at one much in the same way that Mr. Carker used to flash his teeth. And Mr. John Collier gives his sitter a cheerful slap on the back before he says like a shampoo or a turkish bath. Next man<|endoftext|>"]
# ["<|startofprev|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel. Nor is Mr. Quilter's manner less interesting than his matter. He tells us that at this festive season of the year with Christmas and roast beef looming before us, similarly drawn from eating and its results occur most readily to the mind. He has grave doubts whether Sir Frederick Layton's work is really Greek after all and can discover in it but little of Rocky Ithaca. Lennils, pictures are a sort of upguards and atom paintings and Mason's exquisite itals are as national as a jingo poem. Mr. Berkett Foster's landscapes smile at one much in the same way that Mr. Carker used to flash his teeth. And Mr. John Collier gives his sitter a cheerful slap on the back before he says like a shampoo or a turkish bath. Next man<|startoftranscript|><|en|><|transcribe|><|notimestamps|> it is obviously unnecessary for us to point out how luminous these criticisms are, how delicate and expression.<|endoftext|>"]
# ["<|startofprev|> middle classes, and we are glad to welcome his gospel. Nor is Mr. Quilter's manner less interesting than his matter. He tells us that at this festive season of the year with Christmas and roast beef looming before us, similarly drawn from eating and its results occur most readily to the mind. He has grave doubts whether Sir Frederick Layton's work is really Greek after all and can discover in it but little of Rocky Ithaca. Lennils, pictures are a sort of upguards and atom paintings and Mason's exquisite itals are as national as a jingo poem. Mr. Berkett Foster's landscapes smile at one much in the same way that Mr. Carker used to flash his teeth. And Mr. John Collier gives his sitter a cheerful slap on the back before he says like a shampoo or a turkish bath. Next man it is obviously unnecessary for us to point out how luminous these criticisms are, how delicate and expression.<|startoftranscript|><|en|><|transcribe|><|notimestamps|> On the general principles of art and Mr. Quilter writes with equal lucidity.<|endoftext|>"]

```
<br>   

>The suggested modifications to model.generate() in this PR make it possible to have both initial_prompt and the condition_on_previous_text options as in OpenAI, but it would require the user to write their own processing loop to get the same results as OpenAI.

Aimed to address this with the new sequential loop over chunks of the input
<br>  
Also, there are hacks in a few places related to getting the id of the prompt start token and separating it from the prompt text ids. Would this be something we could add to the model or generation config?

Commit:
flip temp check operand

flozi00 commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1634,6 +1634,9 @@ def _get_forced_decoder_ids_with_prompt(text_prompt_ids: List[int]) -> List[Tupl
             kwargs.get("forced_decoder_ids") or generation_config.forced_decoder_ids
         )
+        if condition_on_previous_text and kwargs.get("temperature", 0) <= 0.5:
+            logger.warning("Overriding `condition_on_previous_text` to be False since `temperature` > 0.5")
```
Should it be >= in condition ?

hollance commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +91,9 @@ def decode(self, *args, **kwargs):
         """
         return self.tokenizer.decode(*args, **kwargs)
+
+    def get_prompt_ids(self, text: str):
+        """Converts prompt text to IDs that can be passed to [`~WhisperForConditionalGeneration.generate`]."""
+        prompt_token_id = self.tokenizer.convert_tokens_to_ids("<|startofprev|>")
+        tokenized_text = self.tokenizer(" " + text.strip(), add_special_tokens=False)["input_ids"] if len(text) else []
+        return [prompt_token_id, *tokenized_text]
```
Rather than returning the `prompt_ids` as a list of integers, it would be preferable to have them as a tensor. But even better, `get_prompt_ids()` should use the `return_tensors` argument just like `tokenizer` does, so that the caller can decide between numpy or torch tensors, or a list of integers.

hollance reviewed (changes requested):
Thanks for working on this feature, @connor-henderson . I think you were very inventive in coming up with a solution that allows us to use `initial_prompt` and `condition_on_previous_text`  as in OpenAI. 😄 

However, your implementation doesn't seem to fit in very well with the current design of Transformers. I'll let my colleagues at HF weigh in too, but it might be better to split this functionality as follows:

1. Add the prompt_ids to `model.generate()` as in your earlier version of the PR. All this does is insert the prompt in the `<|startofprev|>` section. This doesn't give us the OpenAI functionality yet, it only adds `<|startofprev|>` support to the modeling and tokenizer code.

2. Create a new pipeline that is specific to Whisper that works more like the OpenAI inference code does. The logic for managing the `<|startofprev|>` section then sits in the new pipeline's loop, not in the model.

(Perhaps step 2 could be a separate PR, to keep the complexity of these PRs down a bit.)

amyeroberts commented:
cc'ing in @gante re `generate`

Commit:
refocusing to prev pr scope

connor-henderson commented:
>1. Add the prompt_ids to model.generate() as in your earlier version of the PR. All this does is insert the prompt in the <|startofprev|> section. This doesn't give us the OpenAI functionality yet, it only adds <|startofprev|> support to the modeling and tokenizer code.

Thanks @hollance I definitely agree splitting this into >1 PR is ideal, have pushed back up code for number 1 above so this can just address that portion. It now implicitly does `always_use_initial_prompt`.

connor-henderson commented:
Curious if by adding `return_tensors` to `get_prompt_ids` you're setting up effectively doing `condition_on_previous_text` via cleverly feeding batches / prompts to `model.generate()` calls (i.e. the first chunk of a second model.generate call would use the text from the first chunk of the first model.generate call as a prompt and so on for each chunk in the batch), but that's more of a question for subsequent PRs

hollance commented:
The reason I asked for the `return_tensors` argument is that passing the `prompt_ids` into `model.generate()` as a `torch.LongTensor` instead of `List[int]` is more consistent with how we normally pass tokens into Transformers models. I understand that inside the model you might need turn it into a list anyway for the `forced_decoder_ids`, but that's really an internal implementation detail. When we generate, the output token sequence is also a Tensor, and so we can concat this to the previous `prompt_ids` to create the next one, etc. I hope that makes sense. :-)

Commit:
remove the params too

Commit:
make style

Commit:
edits, move max length incorporating prompt to whisper

gante commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1604,12 +1616,40 @@ def generate(
+                kwargs.pop("forced_decoder_ids", None) or generation_config.forced_decoder_ids
+            )
+            forced_decoder_ids = [
+                *text_prompt_ids[-self.config.max_length // 2 - 1 :],
```
Is there a reason behind this slicing? Intuitively it makes sense to me, but I'm curious to know if there is a reference behind this choice :)

connor-henderson replied:
Sure I'll leave a comment in the code too, this is done to match Whisper's [implementation](https://github.com/openai/whisper/blob/c09a7ae299c4c34c5839a76380ae407e7d785914/whisper/decoding.py#L586-L601). I believe the reason they do the -1 is to make room for the first token to generate, and the reason they do // 2 is to halve it to share context space with a prefix if one is provided (which also gets halved). I don't believe there's prefix support yet in `transformers` so technically the // 2 isn't necessary at this point but I didn't want to confuse future work around that if it happens. There's a good clarification of prompt vs prefix [here](https://github.com/openai/whisper/discussions/117) if it's of interest.

Helene-Maxcici replied:
Hello @connor-henderson, as I am using the prompting feature I noticed a bug for long prompts. It might be caused by the slicing, where it should be `text_prompt_ids = text_prompt_ids[-(self.config.max_length // 2 - 1) :]`, to correctly account for the first token `<|startofprev|>`.

sanchit-gandhi replied:
Hey @Helene-Maxcici, feel free to open a new issue to track this bug, tagging myself (and optionally @connor-henderson). In particular, it would be super helpful to have a reproducible code snippet to emulate the behaviour locally. See the following page for details: https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#submitting-a-bug-related-issue-or-feature-request

gante commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -89,6 +89,14 @@ def prepare_whisper_inputs_dict(
     }
 
 
+def load_datasamples(num_samples):
```
Ideally, this sort of cosmetic change should be avoided unless there is a strong reason for it. It pollutes the diff and makes tracking changes harder :)

connor-henderson replied:
got it, will remove

gante reviewed (approved):
The PR LGTM as it is, thank you for the contribution 🙌 

BTW, the code changes do not match the description at the top. From what I gathered in the comments, there will be a follow-up PR, correct? In that case, would you mind updating the PR, before I tag a core maintainer for a final review? :)

hollance commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1604,12 +1616,40 @@ def generate(
+            generation_config.max_new_tokens = non_prompt_max_length + len(text_prompt_ids)
+
+            # Reformat the forced_decoder_ids to incorporate the prompt
+            non_prompt_forced_decoder_ids = (
+                kwargs.pop("forced_decoder_ids", None) or generation_config.forced_decoder_ids
+            )
```
Nice! I think this now supports the different ways that `forced_decoder_ids` may be passed in?

1. Through `model.config.forced_decoder_ids = processor.get_decoder_prompt_ids(language=..., task=...)`

2. Through `model.generate(input_features, forced_decoder_ids=forced_decoder_ids)`

3. Through `model.generate(input_features, language=..., task=...)`

It would be good if there are unit tests for these different methods.

connor-henderson replied:
I don't believe `model.generate` allows passing in task or language directly as in 3. above, but I've added tests for the other two

hollance replied:
It does allow that (and I think it might even be the preferred method now) but for some reason the language needs to be the token, such as `"<|ja|>"` rather than `"ja"`.

sanchit-gandhi replied:
Note that the language code change was @connor-henderson's most recent PR! This `forced_generation_ids` logic is in-place so that the code is backward compatible with our previous way of handling the langauge/task, where we either set it in the config as `config.forced_decoder_ids`, or explicitly as `forced_decoder_ids` to the generate method (see https://github.com/huggingface/transformers/pull/21965#discussion_r1127907439)

gante replied:
haha derp, I didn't look at the author 🙈 my bad!

hollance commented on src/transformers/models/whisper/tokenization_whisper_fast.py:
```diff
@@ -296,6 +296,17 @@ def decode(
         Returns:
             `str`: The decoded sentence.
         """
+        token_ids = to_py_obj(token_ids)
+        prompt_token_id = self.convert_tokens_to_ids("<|startofprev|>")
+        has_prompt = isinstance(token_ids, list) and prompt_token_id in token_ids
+        # If an initial prompt was used, we need to remove it when skipping special tokens
+        if has_prompt and skip_special_tokens:
+            for i in range(1, len(token_ids)):
+                prompt_end_idx = i
+                if token_ids[i] in self.all_special_ids:
+                    break
+            token_ids = token_ids[prompt_end_idx:]
+
```
Since this is the same logic as in the regular tokenizer, maybe we can extract it into a shared helper function?

connor-henderson replied:
moved some of the functionality into a helper and left some for what I think is the right reusability/readability tradeoff, lmk if you think more should be abstracted

hollance commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -998,20 +1006,13 @@ class WhisperModelIntegrationTests(unittest.TestCase):
         model = WhisperModel.from_pretrained("openai/whisper-tiny")
         model.to(torch_device)
-        input_speech = self._load_datasamples(1)
+        input_speech = load_datasamples(1)
```
I can understand putting this into a free function but if it's only used in one class, we generally keep it as a member function.

connor-henderson replied:
for sure, removed this change. i'd been moving the tests around and at one point I had two classes using this

hollance commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -1406,6 +1407,23 @@ def test_tiny_specaugment_librispeech(self):
         # fmt: on
         self.assertTrue(torch.allclose(logits[0][0, 0, :30].cpu(), EXPECTED_LOGITS, atol=1e-4))
 
+    @slow
+    def test_generate_with_prompt_ids(self):
+        processor = WhisperProcessor.from_pretrained("openai/whisper-tiny")
+        model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")
+        model.to(torch_device)
+        input_speech = load_datasamples(4)[-1:]
+        input_features = processor(input_speech, return_tensors="pt").input_features
+
+        output_without_prompt = model.generate(input_features)
+        prompt_ids = processor.get_prompt_ids("Leighton")
+        output_with_prompt = model.generate(input_features, prompt_ids=prompt_ids)
```
Nice test! I'd like to see a few more tests where you also change the `forced_decoder_ids` (see my comment above). The way the `forced_decoder_ids` get passed around is a bit brittle (due to the code for that changing a few times) and so we should make sure we have solid tests, since it's too easy for someone to change how this works and inadvertently break something.

hollance commented on tests/models/whisper/test_processor_whisper.py:
```diff
@@ -146,3 +146,11 @@ def test_get_decoder_prompt_ids(self):
         expected_ids = [TRANSCRIBE, NOTIMESTAMPS]
         self.assertListEqual([ids[-1] for ids in forced_decoder_ids], expected_ids)
+
+    def test_get_prompt_ids(self):
```
Could you also add some tests for edge cases?

For example: `processor.get_prompt_ids("")` or `processor.get_prompt_ids("<|startofprev|> Mr. <|startofprev|> Quilter")`

connor-henderson replied:
The second will definitely confuse the model and decoding if they were passed to the current `get_prompt_ids` as is, would you prefer we strip the prompt start token or raise an error that it was included? I'll push up a change that strips it and logs a warning for now

hollance replied:
I don't really know what would be the best approach here, was just trying to think of things that might go wrong. ;-) 

Perhaps raising an error on unexpected input is the best choice, but only if it doesn't add a lot of complexity to the code.

connor-henderson replied:
Looks like they have their tiktoken package handle it and it raises an error if any special token is included, so will look to do the same

hollance reviewed (changes requested):
This PR is shaping up nicely, @connor-henderson! I think this PR has the right amount of changes and then we can figure out how to do the sequential generation in a follow-up PR. 

I've added a bunch of remarks and suggestions so we can make this fit as well into Transformers as possible. 😄 

I'd also like to invite my colleagues @sanchit-gandhi and @ArthurZucker to have a look at these changes.

connor-henderson commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -586,6 +586,17 @@ def decode(
         Returns:
             `str`: The decoded sentence.
         """
+        token_ids = to_py_obj(token_ids)
```
Yea I agree, moved this prompt removal code after that `super().decode(...)` call to `_decode` so this conversion isn't necessary

connor-henderson commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1464,6 +1470,7 @@ def generate(
         task=None,
         language=None,
         is_multilingual=None,
+        prompt_ids: Optional[Union[List[int], torch.Tensor, np.ndarray]] = None,
```
Changed to this suggestion

Commit:
address comments

connor-henderson commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -1406,6 +1406,56 @@ def test_tiny_specaugment_librispeech(self):
+        self.assertEqual(processor.decode(output_with_prompt[0]), expected_with_prompt)
+
+    @slow
+    def test_generate_with_prompt_ids_and_forced_decoder_ids(self):
```
I feel like this is most readable / simplest as one test with comments clarifying the cases, lmk if you want them split into separate unit tests

hollance replied:
I agree it's very readable but there's a potential issue: the `model` will keep state around, i.e. it will change the `model.generation_config` object with the new `forced_decoder_ids` and this may affect the next test. So I think it's better to instantiate a new model before each test.

Maybe it's also a good idea to test what happens after you do the following, just to make sure the code can handle both of these things being None:

```python
model.config.forced_decoder_ids = None
model.generation_config.forced_decoder_ids = None
```

connor-henderson replied:
Added a case for the above which involved a change in `generate` I'll call out below. I was aiming to order the tests to prevent conflicting state issues but you're right they're more brittle that way, split them into individual tests

connor-henderson replied:
Sorry, moving parts, we had a test explicitly for this when there we 5 test cases. Then we trimmed them, per this https://github.com/huggingface/transformers/pull/22496#discussion_r1188856955 I changed the `test_generate_with_prompt_ids_and_no_non_prompt_forced_decoder_ids` test to use `whisper-base.en` and `return_timestampt=True`. I just tested it tho and realized that combination didn't actually set those attributes to `None`, so I updated the test to explicity set those two to `None`.

tl;dr it was tested, then wasn't, now is again

Commit:
remove asr pipeline prompt decoding, fix indexing

connor-henderson commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -606,6 +606,10 @@ def _decode(
     ) -> str:
         self._decode_use_source_tokenizer = kwargs.pop("use_source_tokenizer", False)
 
+        has_prompt = isinstance(token_ids, list) and self.convert_tokens_to_ids("<|startofprev|>") in token_ids
```
went back to using 'in' check here instead of indexing to 0 in token_ids since it errors if empty

hollance replied:
The reason I suggested putting it inside a separate `if skip_special_tokens:` is that the `in` operation needs to scan through the list, which is slow, and we can avoid it if `skip_special_tokens` is False.

connor-henderson replied:
good point i'll adapt it to check the index 0

hollance commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1521,6 +1528,11 @@ def generate(
                 find all the possible language tokens in the `model.generation_config.lang_to_id` dictionary.
             is_multilingual (`bool`, *optional*):
                 Whether or not the model is multilingual.
+            prompt_ids (`Optional[Union[List[int], torch.Tensor, np.ndarray]]`, *optional*):
+                List or rank-1 tensor of token IDs created by passing text to [`~WhisperProcessor.get_prompt_ids`] that
```
Doc comment still refers to the old type hints.

hollance commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +91,12 @@ def decode(self, *args, **kwargs):
+
+    def get_prompt_ids(self, text: str):
+        """Converts prompt text to IDs that can be passed to [`~WhisperForConditionalGeneration.generate`]."""
+        prompt_token = "<|startofprev|>"
+        if prompt_token in text:
+            text = text.replace(prompt_token, "")
```
Would it be better to strip out all special tokens / raise an error? 

Not sure what OpenAI does here. In "condition on previous text mode" they don't include the `<|startoftranscript|><|en|><|transcribe|><|notimestamps|>` tokens when they put the previous text in the `<|startofprev|>` section (since that would be problematic). But I'm not sure if they also strip out the actual timestamps such as `<|1.5|>` etc, that would be worth looking into.

I think `get_prompt_ids` should not accept any tokens `>= processor.tokenizer.eos_token`, so we should either strip these out or raise an error. If we do want to allow timestamp tokens, then it should accept tokens `> processor.tokenizer.all_special_ids[-1]`, since that's where the timestamp tokens begin.

connor-henderson replied:
Looks like they have their tiktoken package handle it and it raises an error if any special token is included, so will look to do the same

connor-henderson replied:
[These](https://github.com/openai/whisper/blob/c09a7ae299c4c34c5839a76380ae407e7d785914/whisper/tokenizer.py#L334-L345) are the tokens they strip out, so timestamps are included. `transformers` uses the same time_precision of 0.02, but notably even tho <|1.00|> is caught by OpenAI's special tokens check any number that doesn't have hundreths place precision like <|1.0|> isn't. I'd guess we still want to raise an error for these so I'll try to write a check for a token with any fp number inside

hollance replied:
Your solution is probably fine but a simpler approach would be to make sure no token has a value `>= processor.tokenizer.eos_token`. ;-)

connor-henderson replied:
Oh interesting, for timestamps and <|nospeech|> too? I get several low ids when trying to tokenize a timestamp

![Screenshot 2023-05-04 at 9 35 47 AM](https://user-images.githubusercontent.com/78612354/236221427-ebd82ee5-3584-4422-b771-788ffedebea6.png)

versae replied:
Maybe related: [#20225](https://github.com/huggingface/transformers/issues/20225#issuecomment-1534911055).

hollance replied:
I guess our tokenizer only decodes timestamp tokens, but doesn't know how to encode them?

Timestamps start at 50364. Any token id higher than that is a timestamp token.

hollance commented on src/transformers/models/whisper/tokenization_whisper_fast.py:
```diff
@@ -312,6 +312,12 @@ def decode(
+        token_ids = kwargs["token_ids"]
+        skip_special_tokens = kwargs["skip_special_tokens"]
+        has_prompt = isinstance(token_ids, list) and token_ids[0] == self.convert_tokens_to_ids("<|startofprev|>")
+        if skip_special_tokens and has_prompt:
```
Same here. I'd prefer to do the `has_prompt` check after checking for `skip_special_tokens`. (It's only a small thing and `skip_special_tokens` will be True most of the time anyway, so consider this a nitpick. ;-) )

hollance reviewed (changes requested):
We're slowly getting there. :smile: (The reason I'm being so nitpicky is that we're going to be changing a model that's already being used a lot, so we have to be very careful to make the right decisions.)

Commit:
address comments (more tests, validate prompt)

Commit:
un-comment out tests (from debug)

connor-henderson commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1566,8 +1577,19 @@ def generate(
+        ):
+            forced_decoder_ids = self.generation_config.forced_decoder_ids
+
+        if task is not None or language is not None or (forced_decoder_ids is None and prompt_ids is not None):
```
This is done solely for handling the case where `prompt_ids` are passed in but the generation config and model config's forced decoder ids are both `None`. Its essentially just changing the order of operations so that we can cleanly check `forced_decoder_ids is None and prompt_ids is not None` to then add non-prompt forced decoder ids, none of the other functionality should change

Commit:
remove old comment

hollance commented:
All right, I think this all looks very good. Pinging @sanchit-gandhi for an additional review since he opened the issue.

sanchit-gandhi commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1521,6 +1527,11 @@ def generate(
             is_multilingual (`bool`, *optional*):
                 Whether or not the model is multilingual.
+            prompt_ids (`Optional[torch.Tensor]]`, *optional*):
+                Rank-1 tensor of token IDs created by passing text to [`~WhisperProcessor.get_prompt_ids`] that is
```
Nice!

sanchit-gandhi commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1595,21 +1617,42 @@ def generate(
+            # Set the decoder_start_token_id to <|startofprev|>
+            kwargs.update({"decoder_start_token_id": decoder_start_token_id})
+
+            # Update the max generation length to include the prompt
```
Well handled!

sanchit-gandhi commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1595,21 +1617,42 @@ def generate(
+                # to accomodate context space for the prefix (see https://github.com/openai/whisper/blob/c09a7ae299c4c34c5839a76380ae407e7d785914/whisper/decoding.py#L599)
+                *text_prompt_ids[-self.config.max_length // 2 - 1 :],
+                generation_config.decoder_start_token_id,
+                *[token for _rank, token in non_prompt_forced_decoder_ids],
```
Is this going to work if `non_prompt_forced_decoder_ids` is `None`? Note that this is only the case for English-only checkpoints that are operating with timestamp prediction. Maybe we can change one of the tests to use `openai/whisper-tiny.en` with `return_timestamp=True` to check this case?

connor-henderson replied:
With the modification mentioned in https://github.com/huggingface/transformers/pull/22496#discussion_r1184394587, `non_prompt_forced_decoder_ids` should never be `None` when `prompt_ids` are passed in. Will have one of the tests cover this

sanchit-gandhi replied:
Great, sounds good to me!

sanchit-gandhi commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +96,22 @@ def decode(self, *args, **kwargs):
         """
         return self.tokenizer.decode(*args, **kwargs)
+
+    def get_prompt_ids(self, text: str):
```
IMO this method should live in the `tokenizer` since it's entirely related to encoding a string to tokens. We can then forward it onto the processor if we need access to it here:
```python
    def get_prompt_ids(self, text: str):
        return self.tokenizer.get_prompt_ids(text)
```

connor-henderson replied:
Moved and kept the tests in the processing test file since that's what was done for `get_decoder_prompt_ids`

sanchit-gandhi commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +96,22 @@ def decode(self, *args, **kwargs):
+        _verify_no_special_tokens(text, self.tokenizer.all_special_tokens)
+        prompt_token = "<|startofprev|>"
+        batch_encoding = self.tokenizer(prompt_token, text.strip(), add_prefix_space=True, add_special_tokens=False)
+        batch_encoding.convert_to_tensors(tensor_type="pt")
```
To make this method compatible with TF and JAX in the future, could we add an argument `return_tensors` to this method and set the tensor type based on that? For consistency with the library, the argument `return_tensors` should default to `np` (NumPy).

sanchit-gandhi commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +96,22 @@ def decode(self, *args, **kwargs):
+    def get_prompt_ids(self, text: str):
+        """Converts prompt text to IDs that can be passed to [`~WhisperForConditionalGeneration.generate`]."""
+        _verify_no_special_tokens(text, self.tokenizer.all_special_tokens)
+        prompt_token = "<|startofprev|>"
```
This can be defined nicely in the `tokenizer` file once we move the method there!

sanchit-gandhi commented on src/transformers/models/whisper/processing_whisper.py:
```diff
@@ -91,3 +96,22 @@ def decode(self, *args, **kwargs):
+        return batch_encoding["input_ids"]
+
+
+def _verify_no_special_tokens(text: str, all_special_tokens: List[str]):
```
Agree with @hollance that checking for any tokens > 50364 is probably cleaner and faster here if we can manage it https://github.com/huggingface/transformers/pull/22496#discussion_r1185251784

connor-henderson replied:
Do you mean 50364 hard coded or the min of the special ids? I'm thinking it's the latter since token-id mappings can vary between tokenizers and some special ids are < 50364.

flozi00 replied:
This could crash if someone use custom vocab when hardcoding the index.

hollance replied:
Yeah best not hardcode it but use `all_special_ids[-1] + 1` which is the first timestamp token.

connor-henderson replied:
Also per @versae 's linked issue above timestamps aren't properly tokenized right now so won't be caught by this check

versae replied:
Couldn't we just add the 1500 timestamps tokens to the tokenizer? That would be in sync with the existing OpenAI implementation. Or maybe add them to the `WhisperTokenizer` class as special `AddedToken`'s? It would mean changing statements like ` all_special_ids[-1] + 1` into  `tokenizer.convert_tokens_to_ids("<|0.00|>")`, but would also enable timestamps encoding out-of-the-box without impacting decoding.

A bit off-topic, but changing the tokenizer to encode the timestamps tokens so training can take those into account is kinda hard, as it touches the code in tons of places (think the trie that is built internally).

sanchit-gandhi commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -988,3 +992,12 @@ def _find_longest_common_sequence(sequences):
+    has_prompt = isinstance(token_ids, list) and token_ids and token_ids[0] == prompt_token_id
+    if has_prompt:
+        for i in range(1, len(token_ids)):
+            if token_ids[i] in all_special_ids:
```
So we just clip to the first time we hit `<|startoftranscript|>`? Think if we know that `has_prompt` is True then we can return the token ids from `<|startoftranscript|>` onwards explicitly (think this will make the code easier to understand)

connor-henderson replied:
Much simpler! It's now short enough I'm wondering if you'd prefer including the below directly in the `_decode` instead of in the static method
```py
if skip_special_tokens:
            decoder_start_token_id = self.convert_tokens_to_ids("<|startoftranscript|>")
            decoder_start_token_id_idx = token_ids.index(decoder_start_token_id)
            token_ids = token_ids[decoder_start_token_id_idx:] # strip prompt if it exists
```

sanchit-gandhi commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -1406,6 +1406,120 @@ def test_tiny_specaugment_librispeech(self):
+        self.assertEqual(processor.decode(output_with_prompt[0]), expected_with_prompt)
+
+    @slow
+    def test_generate_with_prompt_ids_and_forced_decoder_ids_case_one(self):
```
These five test cases are really nice! Probably we can retain just one for the final PR since adding lots of slow tests makes the CI a lot slower

connor-henderson replied:
Added two fast checks, # of slow tests is now 2 tho to accomodate both case one and `openai/whisper-tiny.en` with `return_timestamp=True` if that's alright

sanchit-gandhi replied:
Perfect!

sanchit-gandhi commented on tests/models/whisper/test_processor_whisper.py:
```diff
@@ -146,3 +148,32 @@ def test_get_decoder_prompt_ids(self):
+        self.assertListEqual(prompt_ids.tolist(), [50360, 220])
+        self.assertEqual(decoded_prompt, "<|startofprev|> ")
+
+    def test_get_prompt_ids_with_special_tokens(self):
```
Very nice!

sanchit-gandhi commented on src/transformers/models/whisper/tokenization_whisper.py:
```diff
@@ -988,3 +992,12 @@ def _find_longest_common_sequence(sequences):
     return total_sequence
+
+
+def _strip_prompt(token_ids: List[int], prompt_token_id: int, all_special_ids: List[int]):
```
Can also make this a static method and move inside the class `WhisperTokenizer`:
```python
@staticmethod
def _strip_prompt(token_ids: List[int], prompt_token_id: int, all_special_ids: List[int]):
```

sanchit-gandhi reviewed (commented):
Awesome @connor-henderson! The modelling code changes are very clean 👌 By-and-large agree with the design decisions you've made as well, just some minor suggestions as comments below

Commit:
address comments

Commit:
fix typo

Commit:
remove timestamp token from test

Commit:
make style

Commit:
cleanup

AvivSham commented:
Is there an estimation of when this branch will be merged?

sanchit-gandhi reviewed (approved):
Thanks for iterating @connor-henderson!

Commit:
copy method to fast tokenizer, set max_new_tokens for test

connor-henderson commented on src/transformers/models/whisper/tokenization_whisper_fast.py:
```diff
@@ -485,3 +490,15 @@ def _decode_asr(self, model_outputs, *, return_timestamps, return_language, time
             return_language=return_language,
             time_precision=time_precision,
         )
+
```
Ah yes, thanks

versae commented on src/transformers/models/whisper/tokenization_whisper_fast.py:
```diff
@@ -485,3 +490,30 @@ def _decode_asr(self, model_outputs, *, return_timestamps, return_language, time
+    # Copied from transformers.models.whisper.tokenization_whisper.WhisperTokenizer.get_prompt_ids
+    def get_prompt_ids(self, text: str, return_tensors="np"):
+        """Converts prompt text to IDs that can be passed to [`~WhisperForConditionalGeneration.generate`]."""
+        batch_encoding = self("<|startofprev|>", text.strip(), add_prefix_space=True, add_special_tokens=False)
```
Quick comment here. By default, `WhisperTokenizerFast` sets `add_prefix_space` to `False`. After being inititalized, the value cannot be changed (see related #17391).

```python
from transformers import WhisperTokenizerFast           
                                                                      
tokenizer = WhisperTokenizerFast.from_pretrained("openai/whisper-tiny")
                                                                      
tokenizer("<|startofprev|>", "test", add_special_tokens=False).input_ids
> [50361, 31636]

tokenizer("<|startofprev|>", "test", add_special_tokens=False, add_prefix_space=True).input_ids   
> TypeError: PreTrainedTokenizerFast._batch_encode_plus() got an unexpected keyword argument 'add_prefix_space'
```

Is it necessary to have `add_prefix_space` be `False` by default? Pinging @hollance and @sanchit-gandhi too.

sanchit-gandhi replied:
Since Whisper always infers an extra space after the `<|startoftranscript|>` / `<|startofprev|>`, I see why it could make sense to change `add_prefix_space` to `True` by default. Probably we should have set it to `True` when we added the model in the first place, but right now to swap it would be a breaking change since we would silently change the behaviour of the tokenizer

I'm not sure there's a clean way we can do this directly? Probably easiest is to instantiate the tokenizer with `add_prefix_space=True` where required:
```python
from transformers import WhisperTokenizerFast

tokenizer = WhisperTokenizerFast.from_pretrained("openai/whisper-tiny", add_prefix_space=True)
```

versae replied:
For training with initial prompts, right now we are using two tokenizers, which is not ideal.

sanchit-gandhi commented:
Thanks for the latest round of changes @connor-henderson! Kindly requesting a final review from @amyeroberts!

hollance commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1464,6 +1469,7 @@ def generate(
         task=None,
         language=None,
         is_multilingual=None,
+        prompt_ids: Optional[Union[torch.Tensor, np.ndarray]] = None,
```
Sorry to be a stickler about this one, but why was `np.ndarray` added back in here?

connor-henderson replied:
No worries, I thought it was implied by @sanchit-gandhi 's ask to have the default return_tensors of get_prompt_ids be 'np'. Should I take it out?

hollance replied:
I'll let Sanchit have the final say on this then. 😉

connor-henderson replied:
Got it thanks, put the type back to being just pytorch

Commit:
prompt_ids type just pt

hollance commented:
Since we're all happy with it, I'm pinging @amyeroberts from the core maintainers team to have a final look.

amyeroberts reviewed (commented):
Thanks for adding this! 

Really nice piece of work incl. comments and tests 🤗  I just have one outstanding question/comment before I think we can merge regarding the setting of `generation_config.max_new_tokens`

amyeroberts reviewed (commented):
Thanks for adding this! 

Really nice piece of work incl. comments and tests 🤗  I just have one outstanding question/comment before I think we can merge regarding the setting of `generation_config.max_new_tokens`

Commit:
address Amy's comments

connor-henderson commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1599,21 +1623,42 @@ def generate(
+            specified_max_length = kwargs.pop("max_new_tokens", None) or kwargs.pop("max_length", None)
+            default_max_length = generation_config.max_new_tokens or generation_config.max_length
+            non_prompt_max_length = specified_max_length or default_max_length
+            generation_config.max_new_tokens = non_prompt_max_length + len(text_prompt_ids)
```
Ah yes thanks for catching, changed to update the kwargs with `max_new_tokens` instead of modifying `generation_config` directly so it only updates it after the deepcopy of `generation_config` in `super.generate()`

connor-henderson commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1593,27 +1617,48 @@ def generate(
                     raise ValueError(
                         f"The `{generation_config.task}`task is not supported. The task should be one of `{TASK_IDS}`"
                     )
-            else:
+            elif hasattr(generation_config, "task_to_id"):
```
Thanks @amyeroberts, comments addressed

Only callout I have is that I updated this line to handle the case where an english-only model is being used and the `model.config.forced_decoder_ids` and `generation_config.forced_decoder_ids` are both `None`, since the `generation_config` doesn't have `task_to_id` in that case

Commit:
make style

amyeroberts reviewed (approved):
Awesome work - a great contribution to the library! 🔥 

Thanks for iterating with us on this and for spending so much time to explain in detail thoughts and decisions along the way.

AvivSham commented:
@amyeroberts @connor-henderson 
Hi All, 
Thank you for your great contribution, however I would like a raise a little concern. 
We tried to inference the model using this branch and the latest commit and got some weird results. 
We provide the audio sample in addition to the prompts for easy reproducing:
[WAV file link](https://drive.google.com/file/d/1kbMEuQv8AmTAyJKkARlwx-wfFhI7uilX/view?usp=sharing)

code:
```
from transformers import WhisperForConditionalGeneration, WhisperProcessor
import torchaudio


input_speech, sr = torchaudio.load(
    "sample.wav"
)
model_name = "openai/whisper-medium"
processor = WhisperProcessor.from_pretrained(model_name, cache_dir="artifacts")
model = WhisperForConditionalGeneration.from_pretrained(model_name, cache_dir="artifacts")
input_features = processor(input_speech.squeeze(), sampling_rate=sr, return_tensors="pt").input_features

# --- Without prompt ---
output_without_prompt = model.generate(input_features)
print(processor.decode(output_without_prompt[0], skip_special_tokens=False))
print(processor.decode(output_without_prompt[0], skip_special_tokens=True))

# --- With prompt ---
prompt_ids = processor.get_prompt_ids("Mexico city")
output_with_prompt = model.generate(input_features, prompt_ids=prompt_ids)
print(processor.decode(output_with_prompt[0], skip_special_tokens=False))
print(processor.decode(output_with_prompt[0], skip_special_tokens=True))
```

and this is the trace:
```
<|startoftranscript|><|en|><|transcribe|><|notimestamps|> San Francisco educators. She was teaching in Mexico City.<|endoftext|>
 San Francisco educators. She was teaching in Mexico City.
<|startofprev|> Mexico city<|startoftranscript|><|en|><|transcribe|><|notimestamps|> and<|endoftext|>
 and
```

When we don't pass prompts we get the expected output, but when we do pass prompts (that appear in the transcription) we end up with a bad output.

Note that we did not commit any code changes before running this script.

System:
 - pytorch 2.0.1
 - The test was made on CPU

sanchit-gandhi commented:
Awesome - thanks for the reviews @amyeroberts and @gante, and for the fast iteration and detailed explanations from you @connor-henderson! Excited to see this PR merged when confirmed as ready 🤗

Regarding prompt engineering, my advice would by to try and emulate a full sentence, complete with punctuation and casing, since really what we're providing as the 'prompt' is just the target transcription from a previous window (see https://github.com/openai/whisper/discussions/963#discussioncomment-4987057)

Johnson-NLP commented:
Hi all,
Thanks for the great work on adding prompt in 'model.generate'.
Is it possible to add 'initial_prompt' in the Fine-Tune code with a 'prompt_use_rate'  to control how often to add prompts to the sentences in training sets?
So that we may improve the performance for some special prompts via prompt-tuning.

amyeroberts commented:
@AvivSham Thanks for reporting and @connor-henderson thanks for investigating! 

I think we're good to merge 👍

dgram0 commented:
Thank you so much for adding this! I've found that I occasionally get the following:
`Traceback (most recent call last):
  File "G:\Conda\hfwhisper\lib\site-packages\transformers\models\whisper\modeling_whisper.py", line 1662, in generate
    return super().generate(
  File "G:\Conda\hfwhisper\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "G:\Conda\hfwhisper\lib\site-packages\transformers\generation\utils.py", line 1518, in generate
    return self.greedy_search(
  File "G:\Conda\hfwhisper\lib\site-packages\transformers\generation\utils.py", line 2345, in greedy_search
    next_token_logits = outputs.logits[:, -1, :]
IndexError: index -1 is out of bounds for dimension 1 with size 0`

My workaround is to catch the exception and try again without the prompt_ids.

hollance commented:
Do you have a reproducible example for this @dgram0? That seems like a serious enough bug that needs investigating further.

hollance commented:
@Johnson-NLP 

> Is it possible to add 'initial_prompt' in the Fine-Tune code with a 'prompt_use_rate' to control how often to add prompts to the sentences in training sets?

Sounds like an interesting idea. Would you mind opening a new issue for this? Thanks!

sanchit-gandhi commented:
To get prompting working with fine-tuning, we probably don't want to explicitly add 'prompted' examples per-se, but rather split longer examples up into shorter ones and feed them sequentially through the model, providing previous passages as 'context' to the model.

For example, if we had a training sample that looked like:
```
This is the first sentence. This is the second sentence. And finally, this is the third.
```

Currently what we do is feed it to the model all at once:
```
<|startoftranscript|> This is the first sentence. This is the second sentence. And finally, this is the third. <|endoftranscript|>
```

What we can do is feed the first sentence in:
```
<|startoftranscript|> This is the first sentence. <|endoftranscript|>
```

Then the second sentence, with the first sentence as context:
```
<|startofprev|> This is the first sentence.<|startoftranscript|> This is the second sentence. <|endoftranscript|>
```

And then the third, with both the first and second sentences as context:
```
<|startofprev|> This is the first sentence. This is the second sentence.<|startoftranscript|>  And finally, this is the third.<|endoftranscript|>
```

At inference time, we then just provide the "context" as our prompts:
```
<|startofprev|> This is the prompt<|startoftranscript|> (model generates the rest...)
```

See section 2.3 of the [Whisper paper](https://arxiv.org/pdf/2212.04356.pdf) for an in-depth explanation as to how they achieve this during pre-training. We essentially want to do the same for fine-tuning

dgram0 commented:
> Do you have a reproducible example for this @dgram0? That seems like a serious enough bug that needs investigating further.

I'll try reproducing in a small toy example. It's reproducible on my side with a fine-tuned large private model.

connor-henderson commented:
@dgram0 thanks for sharing, I was able to repro this. As far as its relation to prompting I think this is another case of prompt sensitivity as opposed to a bug, but it may still be of interest with regards to Whisper generally since its the same error message as issue #22682. 

I noticed that joining the prompts by `' - '` was causing the model to start predicting chinese characters, and using `'. '` instead did not lead to the error (at least through 30 loops, at that point I stopped testing). I did notice degraded predictions over time though since a period did not necessarily belong after each result, and every now and again a chinese char was still predicted so. I'd just be cautious about how prompts are chained together.

dgram0 commented:
@connor-henderson It's a bit of a contrived example meant just to recreate the issue without having to loop too much and at the same time show what may be considered a normal use case. Even without it predicting non-English characters or words you'll eventually encounter the issue within a few hundred loops.

dgram0 commented:
> @dgram0 thanks for sharing, I was able to repro this. As far as its relation to prompting I think this is another case of prompt sensitivity as opposed to a bug, but it may still be of interest with regards to Whisper generally since its the same error message as issue #22682.
> 
> I noticed that joining the prompts by `' - '` was causing the model to start predicting chinese characters, and using `'. '` instead did not lead to the error (at least through 30 loops, at that point I stopped testing). I did notice degraded predictions over time though since a period did not necessarily belong after each result, and every now and again a chinese char was still predicted so. I'd just be cautious about how prompts are chained together.

The following still joins the prompts using `' - '`, doesn't allow non-English characters in the prompts, doesn't seem to predict Chinese characters, does a decent job of transcription, and still fails on the 144th loop.

```
from datasets import load_dataset, DatasetDict
from transformers import WhisperForConditionalGeneration, WhisperProcessor
import re
import torch

it = iter(load_dataset("librispeech_asr", "all", split="test.other", streaming=True))
processor = WhisperProcessor.from_pretrained("openai/whisper-tiny", language="English", task="transcribe")
model = WhisperForConditionalGeneration.from_pretrained("openai/whisper-tiny")
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
_ = model.to(device)
prompt = 'Some text rich in domain specific vocabulary and example format lives here.'
past_prompts = ["I am from the cutter lying off the coast."]
while it:
  clip = next(it)
  input_features = processor(clip['audio']['array'], sampling_rate=clip['audio']['sampling_rate'], return_tensors="pt").input_features
  prompt_ids = processor.get_prompt_ids(prompt + ' - ' + ' - '.join(past_prompts))
  if device.type == 'cuda':
    input_features = input_features.cuda()
  pred_ids = model.generate(input_features, language="english", task="transcribe", max_new_tokens=128, prompt_ids=prompt_ids)
  result = processor.batch_decode(pred_ids, skip_special_tokens=True)[0].strip()
  result = re.sub(r"[^\u0000-\u05C0\u2100-\u214F]+$", "", result)
  print(result)
  if result != '':
    past_prompts.append(result)
    if len(past_prompts) > 12:
      past_prompts = past_prompts[1:]
```

connor-henderson commented:
Thanks @dgram0 in that case I think this is a bug, I opened an issue #23723 and PR #23724 for both this and another bug this made me realize where `max_new_tokens` isn't properly enforced when the prompt_ids length is too large. I think they both have the same root cause.

hollance commented:
Thanks, @dgram0. Would you have time to look at this bug @connor-henderson, since you're most familiar with this code? If not, I can have a look.

hollance commented:
@connor-henderson @sanchit-gandhi Hey, did we ever resolve the `add_prefix_space` issue?

If I do the following,

```python
pipe = pipeline(task="automatic-speech-recognition", model="openai/whisper-tiny")
prompt_ids = pipe.tokenizer.get_prompt_ids("Hello, world!", return_tensors="pt")
```

I get the error,

```python
TypeError: _batch_encode_plus() got an unexpected keyword argument 'add_prefix_space'
```

It works fine if I create a `processor` or `tokenizer` object by hand and call `get_prompt_ids()`.

I seem to recall this issue came up before but not sure if anything was decided for it?

connor-henderson commented:
@hollance @versae I missed that just looked into it. Appears to be a difference with the slow tokenizer accepting `add_prefix_space` and the fast tokenizer not recognizing or applying it, opened an issue here: #23764

patrickvonplaten commented on tests/models/whisper/test_modeling_whisper.py:
```diff
@@ -1013,6 +1013,48 @@ def test_mask_time_prob(self):
+        forced_decoder_ids = [(1, 6), (2, 7), (3, 8)]
+
+        output = model.generate(
+            input_features, max_new_tokens=5, forced_decoder_ids=forced_decoder_ids, prompt_ids=prompt_ids
```
Why do we allow passing `prompt_ids` as a numpy array here?

patrickvonplaten commented on src/transformers/models/whisper/modeling_whisper.py:
```diff
@@ -1464,6 +1469,7 @@ def generate(
         task=None,
         language=None,
         is_multilingual=None,
+        prompt_ids: Optional[torch.Tensor] = None,
```
I think `prompt_ids` should not be allowed to be a numpy array given its signature (see: https://github.com/huggingface/transformers/pull/22496/files#r1467369773)
